### PR TITLE
feat(engine): Add support for isConnected on LightningElement

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
Developers can now check if an LWC element is connected to the DOM using the `isConnected` property on the component instance.

Fixes https://github.com/salesforce/lwc/issues/1698

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

## GUS Work Item
W-7199523

### CODE REVIEW

File: packages/@lwc/engine/src/framework/base-lightning-element.ts [hunk 1]
```diff
@@ -487,6 +487,12 @@ BaseLightningElementConstructor.prototype = {
         return elm.getElementsByClassName(names);
     },
 
+    get isConnected(): boolean {
+        const vm = getAssociatedVM(this);
+        const { elm } = vm;
+        return elm.isConnected;
```
Comment (ravijayaramappa) [inline packages/@lwc/engine/src/framework/base-lightning-element.ts]: `isConnected` is polyfiled in @lwc/synthetic-shadow https://github.com/salesforce/lwc/blob/master/packages/@lwc/synthetic-shadow/src/faux-shadow/node.ts#L470

File: packages/@lwc/engine/src/framework/restrictions.ts [hunk 2]
```diff
@@ -432,6 +439,29 @@ function getLightningElementPrototypeRestrictionsDescriptors(
...
             },
         }),
+        isConnected: generateAccessorDescriptor({
+            get(this: LightningElement) {
+                const vm = getAssociatedVM(this);
+                const componentTag = getComponentTag(vm);
+                assert.isFalse(
+                    isBeingConstructed(vm),
+                    `this.isConnected cannot be accessed during the construction phase of the custom` +
+                        ` element for ${componentTag} because it is redundant. The value will always be` +
```
Comment (pmdartus) [inline packages/@lwc/engine/src/framework/restrictions.ts]: Even if the error makes sense in 99% of the case, I would like to bring more nuance to this statement. If the element is already connected to the DOM before the upgrade the `isConnected` will be true in the `constructor`: https://jsfiddle.net/pmdartus/fb1Lt8wa/2/

File: packages/@lwc/engine/src/framework/context.ts [hunk 3]
```diff
@@ -12,9 +12,3 @@ export interface Context {
     tplCache?: Template;
     [key: string]: any;
 }
-
-export let currentContext: Context = {};
```
Comment (pmdartus) [inline packages/@lwc/engine/src/framework/context.ts]: Let's move the context removal to another PR.
